File: src/core/decoder.js
// Raw CAN payload -> signal values.
//
// Bit addressing follows the DBC conventions:
//  - Intel (@1, little-endian): startBit is the position of the LSB; message
//    bit i lives at byte i>>3, bit i&7; significance grows upward.
//  - Motorola (@0, big-endian): startBit is the position of the MSB in
//    msb-first ("sawtooth") numbering; significance decreases within a byte
//    (bit 7 -> 0) and continues at bit 7 of the NEXT byte (+15 in raw bit
//    position). Getting this walk wrong is the classic CAN decoder bug — it is
//    pinned by golden-vector tests in test/.

const f64scratch = new DataView(new ArrayBuffer(8));

/**
 * Extract the raw (unscaled) value of a signal from a payload.
 * Returns a Number (safe: ≤ 52 significant bits handled via BigInt fallback,
 * result converted with Number()). Returns null if the payload is too short.
 */
export function extractRaw(data, signal) {
  const { startBit, bitLength, byteOrder, signed } = signal;

  // Bounds check: highest byte the signal touches.
  const lastBit =
    byteOrder === 1
      ? startBit + bitLength - 1
      : motorolaLastBit(startBit, bitLength);
  if (lastBit >> 3 >= data.length || lastBit < 0) return null;

  let raw;
  if (bitLength <= 32) {
    raw = extract32(data, startBit, bitLength, byteOrder);
    if (signed && bitLength < 32 && raw & (1 << (bitLength - 1))) {
      raw -= 1 << bitLength;
    } else if (signed && bitLength === 32 && raw & 0x80000000) {
      raw -= 0x100000000;
    } else {
      raw = raw >>> 0 === raw ? raw : raw >>> 0; // keep unsigned positive
    }
  } else {
    let big = extractBig(data, startBit, bitLength, byteOrder);
    if (signed && big & (1n << BigInt(bitLength - 1))) {
      big -= 1n << BigInt(bitLength);
    }
    raw = Number(big);
  }

  if (signal.isFloat) {
    // IEEE 754 stored in the extracted bit pattern.
    if (signal.isDouble || bitLength === 64) {
      const big = BigInt.asUintN(64, BigInt(raw < 0 ? raw + 2 ** 64 : raw));
      f64scratch.setBigUint64(0, big);
      return f64scratch.getFloat64(0);
    }
    f64scratch.setUint32(0, raw >>> 0);
    return f64scratch.getFloat32(0);
  }
  return raw;
}

function extract32(data, startBit, bitLength, byteOrder) {
  let value = 0;
  if (byteOrder === 1) {
    for (let i = 0; i < bitLength; i++) {
      const pos = startBit + i;
      value |= ((data[pos >> 3] >> (pos & 7)) & 1) << i;
    }
  } else {
    let pos = startBit;
    for (let i = 0; i < bitLength; i++) {
      value = (value << 1) | ((data[pos >> 3] >> (pos & 7)) & 1);
      pos = (pos & 7) === 0 ? pos + 15 : pos - 1;
    }
  }
  return value >>> 0;
}

function extractBig(data, startBit, bitLength, byteOrder) {
  let value = 0n;
  if (byteOrder === 1) {
    for (let i = 0; i < bitLength; i++) {
      const pos = startBit + i;
      if ((data[pos >> 3] >> (pos & 7)) & 1) value |= 1n << BigInt(i);
    }
  } else {
    let pos = startBit;
    for (let i = 0; i < bitLength; i++) {
      value = (value << 1n) | BigInt((data[pos >> 3] >> (pos & 7)) & 1);
      pos = (pos & 7) === 0 ? pos + 15 : pos - 1;
    }
  }
  return value;
}

function motorolaLastBit(startBit, bitLength) {
  let pos = startBit;
  for (let i = 1; i < bitLength; i++) pos = (pos & 7) === 0 ? pos + 15 : pos - 1;
  return pos;
}

/** raw -> physical engineering value. */
export function rawToPhysical(raw, signal) {
  return raw * signal.factor + signal.offset;
}

/**
 * Decode one signal from a payload to its physical value.
 * Handles multiplexing: returns null when the frame's selector value does not
 * match a multiplexed signal, or when the payload is too short.
 */
export function decodeSignal(data, signal) {
  if (signal.muxRole === 'multiplexed') {
    const mux = signal.message?.multiplexor;
    if (mux) {
      const sel = extractRaw(data, mux);
      if (sel === null || sel !== signal.muxValue) return null;
    }
  }
  const raw = extractRaw(data, signal);
  if (raw === null) return null;
  return rawToPhysical(raw, signal);
}

/** Value-table label for a physical value, or null. */
export function valueLabel(signal, physical) {
  if (!signal.valueTable) return null;
  // VAL_ tables are keyed by raw value.
  const raw = Math.round((physical - signal.offset) / signal.factor);
  return signal.valueTable.get(raw) ?? null;
}

/** Decode every signal of a message payload -> [{signal, raw, value}] (display use). */
export function decodeMessage(data, message) {
  const out = [];
  const mux = message.multiplexor;
  const sel = mux ? extractRaw(data, mux) : null;
  for (const signal of message.signals) {
    if (signal.muxRole === 'multiplexed' && sel !== signal.muxValue) continue;
    const raw = extractRaw(data, signal);
    if (raw === null) continue;
    out.push({ signal, raw, value: rawToPhysical(raw, signal) });
  }
  return out;
}

File: python/canviewer/core/decoder.py
"""Raw CAN payload -> signal values. Port of the JS decoder.

Bit addressing:
 - Intel (@1): startBit is the LSB position; significance grows upward.
 - Motorola (@0): startBit is the MSB position in msb-first ("sawtooth")
   numbering; within a byte significance decreases (bit 7 -> 0) and continues
   at bit 7 of the next byte (+15 in raw bit position).
"""
from __future__ import annotations

import struct


def _motorola_last_bit(start_bit: int, bit_length: int) -> int:
    pos = start_bit
    for _ in range(1, bit_length):
        pos = pos + 15 if (pos & 7) == 0 else pos - 1
    return pos


def extract_raw(data: bytes, signal) -> int | float | None:
    start_bit = signal.start_bit
    bit_length = signal.bit_length
    byte_order = signal.byte_order
    signed = signal.signed

    last_bit = (start_bit + bit_length - 1) if byte_order == 1 else _motorola_last_bit(start_bit, bit_length)
    if last_bit < 0 or (last_bit >> 3) >= len(data):
        return None

    value = 0
    if byte_order == 1:  # Intel / little-endian
        for i in range(bit_length):
            pos = start_bit + i
            value |= ((data[pos >> 3] >> (pos & 7)) & 1) << i
    else:  # Motorola / big-endian sawtooth
        pos = start_bit
        for _ in range(bit_length):
            value = (value << 1) | ((data[pos >> 3] >> (pos & 7)) & 1)
            pos = pos + 15 if (pos & 7) == 0 else pos - 1

    if signal.is_float:
        if signal.is_double or bit_length == 64:
            return struct.unpack("<d", struct.pack("<Q", value & 0xFFFFFFFFFFFFFFFF))[0]
        return struct.unpack("<f", struct.pack("<I", value & 0xFFFFFFFF))[0]

    if signed and (value & (1 << (bit_length - 1))):
        value -= 1 << bit_length
    return value


def raw_to_physical(raw, signal) -> float:
    return raw * signal.factor + signal.offset


def decode_signal(data: bytes, signal):
    if signal.mux_role == "multiplexed":
        mux = signal.message.multiplexor if signal.message else None
        if mux is not None:
            sel = extract_raw(data, mux)
            if sel is None or sel != signal.mux_value:
                return None
    raw = extract_raw(data, signal)
    if raw is None:
        return None
    return raw_to_physical(raw, signal)


def value_label(signal, physical):
    """VAL_ enum/state label for a physical value, or None."""
    if not signal.value_table:
        return None
    raw = round((physical - signal.offset) / signal.factor)
    return signal.value_table.get(raw)


def decode_message(data: bytes, message):
    out = []
    mux = message.multiplexor
    sel = extract_raw(data, mux) if mux else None
    for signal in message.signals:
        if signal.mux_role == "multiplexed" and sel != signal.mux_value:
            continue
        raw = extract_raw(data, signal)
        if raw is None:
            continue
        out.append((signal, raw, raw_to_physical(raw, signal)))
    return out
